### DISCUSSION TITLE
Error building with hypre 2.24.0

### DISCUSSION THREAD

**Discussion by BBBBBbruce** [mfem/mfem#2977] [Q&A] [answered]
Hello, I am trying to build the cuda  parallel version. The packages are mfem-4-3 hypre-2.24.0 metis-5.1.0. 
however I received lots of errors 

`linalg/hypre.cpp:1268:9: error: ‘hypre_ParCSRMatrixOwnsColStarts’ was not declared in this scope; did you mean ‘hypre_ParCSRMatrixColStarts’?
 1268 |         hypre_ParCSRMatrixOwnsColStarts(A)))
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |         hypre_ParCSRMatrixColStarts
linalg/hypre.cpp:1292:37: error: incompatible types in assignment of ‘HYPRE_BigInt*’ {aka ‘int*’} to ‘HYPRE_BigInt [2]’ {aka ‘int [2]’}
 1292 |    hypre_ParCSRMatrixRowStarts(A) = new_row_starts;
      |                                     ^~~~~~~~~~~~~~
linalg/hypre.cpp:1293:4: error: ‘hypre_ParCSRMatrixOwnsRowStarts’ was not declared in this scope; did you mean ‘hypre_ParCSRMatrixRowStarts’?
 1293 |    hypre_ParCSRMatrixOwnsRowStarts(A) = 1;
      |    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |    hypre_ParCSRMatrixRowStarts
linalg/hypre.cpp:1297:40: error: incompatible types in assignment of ‘HYPRE_BigInt*’ {aka ‘int*’} to ‘HYPRE_BigInt [2]’ {aka ‘int [2]’}
 1297 |       hypre_ParCSRMatrixColStarts(A) = new_row_starts;
      |                                        ^~~~~~~~~~~~~~
linalg/hypre.cpp:1298:7: error: ‘hypre_ParCSRMatrixOwnsColStarts’ was not declared in this scope; did you mean ‘hypre_ParCSRMatrixColStarts’?
 1298 |       hypre_ParCSRMatrixOwnsColStarts(A) = 0;
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |       hypre_ParCSRMatrixColStarts
linalg/hypre.cpp: In member function ‘void mfem::HypreParMatrix::CopyColStarts()’:
linalg/hypre.cpp:1304:14: error: ‘hypre_ParCSRMatrixOwnsColStarts’ was not declared in this scope; did you mean ‘hypre_ParCSRMatrixColStarts’?
 1304 |    if (!A || hypre_ParCSRMatrixOwnsColStarts(A) ||
      |              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |              hypre_ParCSRMatrixColStarts
`

Looks like some re-naming issues, and there were too many of them.
These error goes off when I switch to hypre-2.16.0 as same in the instruction. I wonder if there is some features in the later version of hypre is needed? Or 2.16.0 is sufficient?

Thanks

**Comment by BBBBBbruce** [comment]
also, I wonder if there is a tutorial on how can I build mfem as an external library that I can used for my own project? is it enough for me to just in include libmfem.a and the header files?

**Comment by v-dobrev** [comment, marked as answer]
Hi @BBBBBbruce,

With hypre v2.24.0 you should use the newer mfem v4.4.

Regarding using mfem from your project: if you use GNU make, you can include the file `config.mk` (from the directory `config/` in the build-tree or from the directory `share/mfem/` in the install tree) -- this file provides a number of useful definitions (`MFEM_*`) that you can use to build with mfem; for example, see how the makefile in the `examples/` directory use that file. If you are using CMake you can use the file `MFEMConfig.cmake` exported by mfem (in the root directory in the build-tree or in the directory `lib/cmake/mfem/` in the install tree); for example, see how the CMake file in GLVis uses the MFEM exported CMake file: https://github.com/GLVis/glvis/blob/master/CMakeLists.txt.
